clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -201,13 +203,16 @@ public int compare(TxnRequestHandler o1, TxnRequestHandler o2) {
 
     public synchronized TransactionalRequestResult initializeTransactions() {
         ensureTransactional();
-        transitionTo(State.INITIALIZING);
-        setProducerIdAndEpoch(ProducerIdAndEpoch.NONE);
-        this.nextSequence.clear();
-        InitProducerIdRequest.Builder builder = new InitProducerIdRequest.Builder(transactionalId, transactionTimeoutMs);
-        InitProducerIdHandler handler = new InitProducerIdHandler(builder);
-        enqueueRequest(handler);
-        return handler.result;
+        if (transactionalRequestResult == null || transactionalRequestResult.isCompleted()) {
```

Comment:
In the current code, calling `initializeTransactions` more than once causes an illegal state. I think we should preserve that semantic in spite of retries after timeouts. In other words, the only situation where you're allowed to retry a call to `initializeTransactions` is after a timeout. Once it returns successfully for the user, we go back to the current behavior and raise an illegal state.

We're almost there with this patch, but the user could see the illegal state before a successful call because of the `isCompleted` check here. To implement the semantics we want, I think we may have to do the following:

1. Move the caching of the initTransactions result into `KafkaProducer`.
2. On the first invocation, we cache the result.
3. If there is a timeout, we continue waiting on the result.
4. Once the result completes successfully, we should return and set the result to null.

Once the cached result is cleared, if the user tries to `initializeTransactions` again, they will get the current illegal state error, which is what we want.

Reply:
Moving the cache result to `KafkaProducer` has to change the signature of `initTransactions` which might need a KIP to have it discussed. Is that true?

Reply:
What I mean is that we would create a private field for the cached result in `KafkaProducer`, so the signature does not change. The reason I am suggesting we move it there is that that is where we are doing the wait, which means we know when the operation has been successfully completed from the user's perspective.